[LB-1514] Import users' playlists from Spotify

rimma-kubanova opened the pull request:
# Problem

[LB-1514](https://tickets.metabrainz.org/browse/LB-1514)
- There were no features that would have enabled to import a whole playlist from Spotify

(Currently, there is a no ticket regarding that, later I will create it and update the PR)

# Solution

- Added a new button "Import from..." in Playlists pages

# How it works?

- If user is authenticated with spotify, the modal will display user's playlists from Spotify. 
- Once clicked, the playlist will be automatically imported to LB. 
- The fetching playlists, getting tracks and converting them to JSPF format are all done in backend
- The Frontend handles creating a playlist with tracks.

# Demo
<img width="500" alt="Screenshot 2024-02-29 at 14 30 29" src="https://github.com/metabrainz/listenbrainz-server/assets/115300909/5ac7899f-4527-4471-9227-f06235df7413">

This is how the page will look like (the "Import should be on right-top side aligned with the "Playlists" and "Collaborators" buttons)

<img width="500" alt="Screenshot 2024-02-29 at 11 26 32" src="https://github.com/metabrainz/listenbrainz-server/assets/115300909/cfa10a2b-cd90-4c09-8583-9a084c35ccf2">

The modal that will popup, when the Import button is clicked. Here will be presented Playlists names (fetched from Spotify) When clicked they will playlist is automatically imported.

<img width="500" alt="Screenshot 2024-02-29 at 14 30 29" src="https://github.com/metabrainz/listenbrainz-server/assets/115300909/5ac7899f-4527-4471-9227-f06235df7413">

Here I have tested it using my own user-token from LB. Here is the screenshot of playlist imported from Spotify. 

Note: the songs that were not found from Spotify in MB (well for some reason it happens) are automatically ignored

# Things to work on

- Update button style (so I have tried to make it dropdown just like "Add Listen" button, however I had some difficulties with alligning it properly and I gave up. So will try to update it later)
- Add a error toast message if user is not authenticated within Spotify
- Add loading screen. So importing playlists from Spotify takes a quite time (~5-20 sec). So I think to add a load screen so users wouldnt get confused.
- Feedback from Monkey, Aerozol: "need a good way to point users that aren't connected to Spotify to [the relevant page](https://listenbrainz.org/settings/music-services/details/)."

pep8speaks commented:
Hello @rimma-kubanova! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`listenbrainz/troi/export.py`](https://github.com/metabrainz/listenbrainz-server/blob/93452bb4302e8ccbaa7f2f6acdb5bf74bbc50e67/listenbrainz/troi/export.py):

> [Line 25:48](https://github.com/metabrainz/listenbrainz-server/blob/93452bb4302e8ccbaa7f2f6acdb5bf74bbc50e67/listenbrainz/troi/export.py#L25): [W292](https://duckduckgo.com/?q=pep8%20W292) no newline at end of file

* In the file [`listenbrainz/webserver/views/playlist_api.py`](https://github.com/metabrainz/listenbrainz-server/blob/93452bb4302e8ccbaa7f2f6acdb5bf74bbc50e67/listenbrainz/webserver/views/playlist_api.py):

> [Line 14:120](https://github.com/metabrainz/listenbrainz-server/blob/93452bb4302e8ccbaa7f2f6acdb5bf74bbc50e67/listenbrainz/webserver/views/playlist_api.py#L14): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 14:131](https://github.com/metabrainz/listenbrainz-server/blob/93452bb4302e8ccbaa7f2f6acdb5bf74bbc50e67/listenbrainz/webserver/views/playlist_api.py#L14): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (134 > 130 characters)
> [Line 829:1](https://github.com/metabrainz/listenbrainz-server/blob/93452bb4302e8ccbaa7f2f6acdb5bf74bbc50e67/listenbrainz/webserver/views/playlist_api.py#L829): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1
> [Line 848:1](https://github.com/metabrainz/listenbrainz-server/blob/93452bb4302e8ccbaa7f2f6acdb5bf74bbc50e67/listenbrainz/webserver/views/playlist_api.py#L848): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 851:67](https://github.com/metabrainz/listenbrainz-server/blob/93452bb4302e8ccbaa7f2f6acdb5bf74bbc50e67/listenbrainz/webserver/views/playlist_api.py#L851): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 852:1](https://github.com/metabrainz/listenbrainz-server/blob/93452bb4302e8ccbaa7f2f6acdb5bf74bbc50e67/listenbrainz/webserver/views/playlist_api.py#L852): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 857:1](https://github.com/metabrainz/listenbrainz-server/blob/93452bb4302e8ccbaa7f2f6acdb5bf74bbc50e67/listenbrainz/webserver/views/playlist_api.py#L857): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 858:1](https://github.com/metabrainz/listenbrainz-server/blob/93452bb4302e8ccbaa7f2f6acdb5bf74bbc50e67/listenbrainz/webserver/views/playlist_api.py#L858): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1
> [Line 877:1](https://github.com/metabrainz/listenbrainz-server/blob/93452bb4302e8ccbaa7f2f6acdb5bf74bbc50e67/listenbrainz/webserver/views/playlist_api.py#L877): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 879:92](https://github.com/metabrainz/listenbrainz-server/blob/93452bb4302e8ccbaa7f2f6acdb5bf74bbc50e67/listenbrainz/webserver/views/playlist_api.py#L879): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 880:1](https://github.com/metabrainz/listenbrainz-server/blob/93452bb4302e8ccbaa7f2f6acdb5bf74bbc50e67/listenbrainz/webserver/views/playlist_api.py#L880): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 891:58](https://github.com/metabrainz/listenbrainz-server/blob/93452bb4302e8ccbaa7f2f6acdb5bf74bbc50e67/listenbrainz/webserver/views/playlist_api.py#L891): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 894:1](https://github.com/metabrainz/listenbrainz-server/blob/93452bb4302e8ccbaa7f2f6acdb5bf74bbc50e67/listenbrainz/webserver/views/playlist_api.py#L894): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 898:1](https://github.com/metabrainz/listenbrainz-server/blob/93452bb4302e8ccbaa7f2f6acdb5bf74bbc50e67/listenbrainz/webserver/views/playlist_api.py#L898): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 907:1](https://github.com/metabrainz/listenbrainz-server/blob/93452bb4302e8ccbaa7f2f6acdb5bf74bbc50e67/listenbrainz/webserver/views/playlist_api.py#L907): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace

* In the file [`listenbrainz/webserver/views/views_utils.py`](https://github.com/metabrainz/listenbrainz-server/blob/93452bb4302e8ccbaa7f2f6acdb5bf74bbc50e67/listenbrainz/webserver/views/views_utils.py):

> [Line 11:26](https://github.com/metabrainz/listenbrainz-server/blob/93452bb4302e8ccbaa7f2f6acdb5bf74bbc50e67/listenbrainz/webserver/views/views_utils.py#L11): [E225](https://duckduckgo.com/?q=pep8%20E225) missing whitespace around operator
> [Line 79:1](https://github.com/metabrainz/listenbrainz-server/blob/93452bb4302e8ccbaa7f2f6acdb5bf74bbc50e67/listenbrainz/webserver/views/views_utils.py#L79): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1
> [Line 86:1](https://github.com/metabrainz/listenbrainz-server/blob/93452bb4302e8ccbaa7f2f6acdb5bf74bbc50e67/listenbrainz/webserver/views/views_utils.py#L86): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1
> [Line 93:1](https://github.com/metabrainz/listenbrainz-server/blob/93452bb4302e8ccbaa7f2f6acdb5bf74bbc50e67/listenbrainz/webserver/views/views_utils.py#L93): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1
> [Line 105:1](https://github.com/metabrainz/listenbrainz-server/blob/93452bb4302e8ccbaa7f2f6acdb5bf74bbc50e67/listenbrainz/webserver/views/views_utils.py#L105): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 106:1](https://github.com/metabrainz/listenbrainz-server/blob/93452bb4302e8ccbaa7f2f6acdb5bf74bbc50e67/listenbrainz/webserver/views/views_utils.py#L106): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1
> [Line 109:16](https://github.com/metabrainz/listenbrainz-server/blob/93452bb4302e8ccbaa7f2f6acdb5bf74bbc50e67/listenbrainz/webserver/views/views_utils.py#L109): [E203](https://duckduckgo.com/?q=pep8%20E203) whitespace before ':'
> [Line 112:19](https://github.com/metabrainz/listenbrainz-server/blob/93452bb4302e8ccbaa7f2f6acdb5bf74bbc50e67/listenbrainz/webserver/views/views_utils.py#L112): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ':'
> [Line 114:23](https://github.com/metabrainz/listenbrainz-server/blob/93452bb4302e8ccbaa7f2f6acdb5bf74bbc50e67/listenbrainz/webserver/views/views_utils.py#L114): [W292](https://duckduckgo.com/?q=pep8%20W292) no newline at end of file

amCap1712 reviewed (changes requested):
Hi! 

Thank you for the PR. I leave reviewing the frontend stuff to @MonkeyDo. As for the backend, the changes look sane but need to be in troi instead. For instance, see how export to spotify (https://github.com/metabrainz/listenbrainz-server/blob/master/listenbrainz/troi/export.py) just collects the auth credentials and calls troi to do all the actual work.

The spotify export code in troi is mostly in https://github.com/metabrainz/troi-recommendation-playground/blob/main/troi/tools/spotify_lookup.py and some additional handling for command line options in cli.py. My suggestion is the move the import code to troi and open a PR there for review.

rimma-kubanova commented:
@amCap1712 Hi, thank you so much for the feedback. I found out from @/mayhem  that this feature was supposed to be in Troi only after committing this PR. No problem, I will migrate the code accordingly!

However, if it will be in Troi, should I use a playlist generator similar to the one used in the export function? Should I create a patch for it? Currently, the code requests an LB API to create a playlist, but I think it might not fit well if the code were in Troi.

P.S. I was thinking of adding this feature to my proposal for GSoC, but I suppose I'll finish it by then.

MonkeyDo commented on frontend/js/src/artist/ArtistPage.tsx:
```diff
@@ -27,7 +27,11 @@ import {
   popularRecordingToListen,
 } from "../album/utils";
-import type { PopularRecording, ReleaseGroup, SimilarArtist } from "../album/utils";
+import type {
```
The are a bunch of files in the PR that only have automatic formatting changes. Ideally those should not be committed as there is no reason to add them to the PR.
If you are doing `git add .` before committing then you need to start reviewing the changes that you actually want to commit :)

rimma-kubanova replied:
Hi! Yeah so sorry for that, I usually remove these files, but forgot about them this time. I will change it back to normal

MonkeyDo commented on frontend/css/playlists.less:
```diff
@@ -71,6 +71,18 @@
   }
 }
 
+.import-playlist-btn {
```
I see you are still working on this as these classes aren't used anywhere.
I made a mockup using the existing .add-listen-btn class which does the same thing as this one (a good sign that we can refactor it to use only one), based on the dashboard page "add listen" button (see code ref below):

![image](https://github.com/metabrainz/listenbrainz-server/assets/6179856/42f03fbb-77bf-4c16-986c-9a0ddbc568b7)

https://github.com/metabrainz/listenbrainz-server/blob/c8b5abf56ba77b3b1c81465b01e51b0e144ac6e1/frontend/js/src/user/Dashboard.tsx#L786-L797

rimma-kubanova replied:
Thank you very much for your feedback! Wow, there's a lot to work on, but I appreciate all the detailed instructions, such a lifesaver!

MonkeyDo reviewed (commented):
Thanks for working on this feature!

I left some preliminary review, but I would like to be able to deploy the PR on our test server to try it out once the backend parts have been moved to troi, and once you have finished the button styling.

rimma-kubanova commented:
Hi! I updated this PR. Here are the changes:

- As @amCap1712 requested, I migrated the code to troi patch. Now the code is in [import_ms.py](https://github.com/metabrainz/listenbrainz-server/blob/dbd1cfe7558d3cd8187478ee86a9e2a6ae69de35/listenbrainz/troi/import_ms.py). 
- The import function now fetches the tracks, gets their mbids and constructs the playlist using PlaylistElement(). Then submits the playlist to Listenbrainz
- As @MonkeyDo suggested, I changed the front a little. I used model.resolve() to get the new playlist and automatically show it in playlist folders. Also, removed console.logs(), my mistake :_)
- Added style for button (but couldn't allign it properly, still figuring out)

MonkeyDo commented on listenbrainz/troi/import_ms.py:
```diff
@@ -0,0 +1,73 @@
+from troi.musicbrainz.recording import RecordingListElement
+import sys
+
+def import_from_spotify(token, user, playlist_id):
```
Not sure exactly what the issue is, but I deployed the PR on test.LB and I couldn't import a playlist.
One issue is clear regarding rate-limiting, considering the flurry of 429 error codes
 
Here are the container logs for one import attempt:
```
Error occurred: 429
Error occurred: 429
Error occurred: 429
Error occurred: 429
Error occurred: 429
Error occurred: 429
Error occurred: 429
Error occurred: 429
Error occurred: 429
Error occurred: 429
Error occurred: 429
Error occurred: 429
Error occurred: 429
Error occurred: 429
Error occurred: 429
Error occurred: 429
Error occurred: 429
Error occurred: 429
Error occurred: 429
Error occurred: 429
Error occurred: 429
Error occurred: 429
Error occurred: 429
Error occurred: 429
Error occurred: 429
Error occurred: 429
Error occurred: 429
Error occurred: 429
Error occurred: 429
Error occurred: 429
Error occurred: 429
Error occurred: 429
Error occurred: 429
Error occurred: 429
Error occurred: 429
Error occurred: 429
Error occurred: 429
Error occurred: 429
Error occurred: 429
Error occurred: 429
Error occurred: 429
Error occurred: 429
Error occurred: 429
Error occurred: 429
Error occurred: 429
Error occurred: 429
Error occurred: 429
Error occurred: 429
Error occurred: 429
Error occurred: 429
Error occurred: 429
Error occurred: 429
Error occurred: 429
Error occurred: 429
Error occurred: 429
Error occurred: 429
Error occurred: 429
Error occurred: 429
Error occurred: 429
Error occurred: 429
Error occurred: 429
Error occurred: 429
Error occurred: 429
Error occurred: 429
Error occurred: 429
Error occurred: 429
Error occurred: 429
Error occurred: 429
Error occurred: 429
Error occurred: 429
2024-03-19 16:56:48,701 listenbrainz.webserver ERROR    Exception on /1/playlist/spotify/6oSNou76m5UwIpe9PH8vm4/tracks [GET]
Traceback (most recent call last):
  File "/usr/local/lib/python3.11/site-packages/flask/app.py", line 1455, in wsgi_app
    response = self.full_dispatch_request()
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/flask/app.py", line 869, in full_dispatch_request
    rv = self.handle_user_exception(e)
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/flask/app.py", line 867, in full_dispatch_request
    rv = self.dispatch_request()
         ^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/flask/app.py", line 852, in dispatch_request
    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/code/listenbrainz/listenbrainz/webserver/decorators.py", line 18, in decorator
    resp = make_response(f(*args, **kwargs))
                         ^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/brainzutils/ratelimit.py", line 262, in rate_limited
    return f(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^
  File "/code/listenbrainz/listenbrainz/webserver/decorators.py", line 42, in decorator
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/code/listenbrainz/listenbrainz/webserver/views/playlist_api.py", line 902, in import_tracks_from_spotify_to_playlist
    playlist = import_from_spotify(token["access_token"], user["auth_token"], playlist_id)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/code/listenbrainz/listenbrainz/troi/import_ms.py", line 26, in import_from_spotify
    recordings = [Recording(mbid=track["recording_mbid"]) for track in mbid_mapped_tracks if "recording_mbid" in track]
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/code/listenbrainz/listenbrainz/troi/import_ms.py", line 26, in <listcomp>
    recordings = [Recording(mbid=track["recording_mbid"]) for track in mbid_mapped_tracks if "recording_mbid" in track]
                                                                                             ^^^^^^^^^^^^^^^^^^^^^^^^^
TypeError: argument of type 'NoneType' is not iterable

```

amCap1712 replied:
Hmm, its possible that the endpoint is returning None for tracks which couldn't be found in the mapping/cache.

MonkeyDo commented on .vscode/settings.json:
```diff
@@ -2,5 +2,8 @@
-}
+  "python.formatting.provider": "none",
+  
+  "editor.codeActionsOnSave": { "source.fixAll": true },
```
Probably shouldn't have committed this file, unless it was on purpose?

MonkeyDo commented on frontend/css/playlists.less:
```diff
@@ -71,6 +71,18 @@
+  margin-left: auto;
+}
+
+.playlist-header {
+  margin-top: 1em;
+  display: flex;
+  justify-content: space-between;
+}
+
```
Probably don't need this class.
Regarding the import button placement, the only thing you need is the `tertiary-nav` css class on the parent component on [line 136](https://github.com/metabrainz/listenbrainz-server/pull/2800/files#diff-50740f3f836f659fc01031697d7a52c784011746a687be5433890665a4d76cb1R136) and you can remove the `style={{ marginTop: "1em" }}`.
Result:
![image](https://github.com/metabrainz/listenbrainz-server/assets/6179856/9e0dfc47-47a4-463d-a7da-513b17da810f)

MonkeyDo commented on frontend/js/src/user/playlists/components/ImportPlaylistModal.tsx:
```diff
@@ -0,0 +1,169 @@
+      return;
+    }
+
+    try {
```
One thing we are missing here is a loading indicator,
Since the import can take some time, it would be useful for users to see that something is going on after they clicked on a playlist name.

You can store a list of playlist IDs in state (say `playlistsSelectedForImport` or something).
Then you could show an indicator directly on the item in the list of playlists, which will make it clear after clicking that import is in progress.

After a successful or failed playlist import, don't forget to remove said playlist ID from the (playlistsSelectedForImport) state to remove the indicator.

MonkeyDo commented on frontend/js/src/user/playlists/components/ImportPlaylistModal.tsx:
```diff
@@ -0,0 +1,169 @@
+  const [playlists, setPlaylists] = React.useState<SpotifyPlaylistObject[]>([]);
+
+  React.useEffect(() => {
+    async function importPlaylistFromSpotify() {
+      try {
+        const response = await APIService.importPlaylistToSpotify(
```
Naming here can be improved. You're not importing the playlists per se.
Maybe `getUserPlaylistsFromSpotify` or something like that ?

MonkeyDo commented on frontend/js/src/user/playlists/components/ImportPlaylistModal.tsx:
```diff
@@ -0,0 +1,169 @@
+    }
+
+    try {
+      const newPlaylist: JSPFPlaylist = await APIService.getSpotifyTracksInJSPF(
```
Similarly here the naming is a bit confusing, maybe left over from the previous implementation.
Maybe simply `importSpotifyPlaylist`?

MonkeyDo reviewed (changes requested):
I was able to import a short playlist successfully, and it worked pretty well !
However longer playlists are throwing errors for me (see comment below).

amCap1712 commented on listenbrainz/troi/import_ms.py:
```diff
@@ -0,0 +1,73 @@
+from troi.musicbrainz.recording import RecordingListElement
+import sys
+
+def import_from_spotify(token, user, playlist_id):
```
We should move this import feature to troi itself and then call troi here.

rimma-kubanova commented:
@MonkeyDo Thanks for the awesome review! I have worked on every point you mentioned and could actually improve the look. Now, when a playlist is selected, a loading circle is displayed indicating that the import is in progress. As you suggested I made it that way, so for each playlist, there would be its own loading circle indicating it. 

Also, finally could align the button as needed.

However, I could very well fix the rate limiting issue. It comes from troi's patch and I'm guessing it happens when too many playlists are selected at once. However, this bug is very random and in most cases I haven't been able to recreate it.
In good news, I have resolved the NoneType error issue on the backend.

@amCap1712 I wanted to ask, you said that this function needs to be in troi itself, could you please help with this one? To what folder I should specifically move this feature? Should I commit it in troi patch github?

MonkeyDo commented on frontend/js/src/user/playlists/Playlists.tsx:
```diff
@@ -23,6 +23,7 @@ import CreateOrEditPlaylistModal from "../../playlists/components/CreateOrEditPl
 import ImportPlaylistModal from "../../playlists/components/ImportPlaylistModal";
 import PlaylistsList from "./components/PlaylistsList";
 import { getPlaylistId, PlaylistType } from "../../playlists/utils";
+import ImportPlaylistModal from "./components/ImportPlaylistModal";
```
There's a duplicate import of `ImportPlaylistModal` above that should be removed (or this line)

MonkeyDo replied:
Neeever mind, I realize this is due to a small conflict with #2805 that adds another button for importing playlist JSPF files.

I hope you don't mind, I'm going to resolve the issue now because otherwise I can't build the image to deploy it and test,

MonkeyDo commented on frontend/js/src/user/playlists/components/ImportPlaylistModal.tsx:
```diff
@@ -0,0 +1,190 @@
+    }
+
+    if (loading.includes(playlistID)) {
+      setLoading(loading.filter((i: string) => i !== playlistID));
```
Why remove the playlist from the loading array here?

MonkeyDo commented on frontend/js/src/user/playlists/components/ImportPlaylistModal.tsx:
```diff
@@ -0,0 +1,190 @@
+        { toastId: "create-playlist-success" }
+      );
+      setLoading(loading.filter((i: string) => i !== playlistID));
+      modal.resolve(newPlaylist);
```
I guess this means we're closing the modal and updating the playlist page with that single playlist after the  import finishes, which is maybe not what you were planning on?

Not sure what to do about this then. We could store an array of successfully imported playlists and send that back once the user click on a "done" button for example, which would require a change in `onPlaylistCreated` to accept an array of JSPFPlaylist

rimma-kubanova replied:
Oh man, I totally forgot about implementing that part. 

Im thinking, since this function of importing playlists are called by each playlist button separately, each of them can call the modal.resolve and pass the playlist the created (or this is not how it works, srry if im mistaken)

MonkeyDo replied:
I am pretty sure if you call .resolve it will resolve the promise that we created by using `NiceModal.show(...).then(...)`, and consequently only the first created playlist will be passed to `onPlaylistCreated`.

In order to resolve multiple playlists, let ;e expand on the solution I suggested above:
1. modify the resolved value function [here](https://github.com/metabrainz/listenbrainz-server/pull/2800/files#diff-50740f3f836f659fc01031697d7a52c784011746a687be5433890665a4d76cb1R181-R185) to accept a JSPFPlaylist or an array of JSPFPlaylist -> `NiceModal.show<JSPFPlaylist | JSPFPlaylist[], any>(...`
2. In the .then function callback just below, check if the `playlist` value is an array (`if(Array.isArray(...))`), and if so send each item in the array to `onPlaylistCreated` 
3.  Modify `ImportSpotifyPlaylistModal` to store successfully created playlists in an array in the component state
4. Create a new function `resolveAndClose` that will do two things: 1. call modal.resolve with the stored playlist array and 2. call the `closeModal` function
5. Make the close button of the modal call `resolveAndClose` instead of closeModal

MonkeyDo reviewed (commented):
Haven't deployed the updated PR to test it, but I have a couple of comments:

rimma-kubanova commented:
Hi @MonkeyDo ! Hope you doing well! I have commited the changes you requested (check [4125ba4](https://github.com/metabrainz/listenbrainz-server/pull/2800/commits/4125ba40019e187e60b89c28daf97362c8211003) ) ! I believe everything is ready now! 

Regarding loading screen, I decided to change it a little bit. Until now, the loading indicator was inside a button with a playlist name, but I believe it doesnt look well and might be confusing. So I made it that way that user can import one playlist at a time, and the loading indicator will be on the bottom . See the screenshot below: (The will be also text saying "Loading {playlistName}. It might take some time")

![image](https://github.com/metabrainz/listenbrainz-server/assets/115300909/4db45710-d145-4402-8230-382c4940a1e6)


If everything is okay, I will switch to fixing test errors !
Thank you for your patience and help !

MonkeyDo commented on listenbrainz/webserver/views/playlist_api.py:
```diff
@@ -11,6 +11,9 @@
 from listenbrainz.troi.export import export_to_spotify
+from listenbrainz.troi.import_ms import import_from_spotify
+from listenbrainz.webserver.views.views_utils import get_user_playlists
+#     get_tracks_from_playlist, mbid_mapping_spotify, listen_to_jspf
```
Some commented imports here; if they are not used here anymore they should be deleted

MonkeyDo commented on frontend/js/src/user/playlists/components/ImportSpotifyPlaylistModal.tsx:
```diff
@@ -0,0 +1,190 @@
+              type="button"
+              className="btn btn-default"
+              data-dismiss="modal"
+              onClick={() => resolveAndClose()}
```
```suggestion
              onClick={resolveAndClose}
```

MonkeyDo commented on frontend/js/src/user/playlists/components/ImportSpotifyPlaylistModal.tsx:
```diff
@@ -0,0 +1,190 @@
+              className="close"
+              data-dismiss="modal"
+              aria-label="Close"
+              onClick={() => resolveAndClose()}
```
```suggestion
              onClick={resolveAndClose}
```

MonkeyDo reviewed (changes requested):
Looking good! We're getting close to being able to merge and use this new feature, I'm looking forward to it !

I think at some point during resolving a merge conflict we lost the `get_user_playlists` method that was exported from the view_utils.py file.


Will deploy to test.listenbrainz.org once that's been added back for more thorough testing

rimma-kubanova commented:
[2798c6e](https://github.com/metabrainz/listenbrainz-server/pull/2800/commits/2798c6edf5c4c0b2a511c9a3e48ade5d9be3d1ab) 
* I decided to remove the get_user_playlists function from view_utils and just transferred its functionality to function in playlist_api.py. I think that way it will be more convenient

* Also fixed the suggestions you requested! 
thank you!

rimma-kubanova commented:
@MonkeyDo I have fixed the problems regarding my code but the tests are failing due to functions in views_utils.py. Not sure how to fix that, since I haven't changed them

MonkeyDo commented on listenbrainz/webserver/views/views_utils.py:
Looks like the tests are failing because of this unintended change, supposedly a merge conflict resolution issue (maybe my fault!)
I'll correct that and tests should pass.
